src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+                {
+                    // Otherwise we just queue any new design time inputs, and still do date checks
+                    foreach (string file in inputs.Inputs.Except(_designTimeInputs, StringComparers.Paths))
+                    {
+                        ImmutableInterlocked.TryAdd(ref _filesToCompile, file, /* ignoreWriteTime */ false);
+                    }
+                }
+            }
+
+            // Make sure we have the up to date list of inputs
+            _designTimeInputs = inputs.Inputs;
+            _sharedDesignTimeInputs = inputs.SharedInputs;
+
+            // Make sure we have the up to date output path
+            string basePath = configChanges.After.Properties[ConfigurationGeneral.ProjectDirProperty];
+            string objPath = configChanges.After.Properties[ConfigurationGeneral.IntermediateOutputPathProperty];
+            _outputPath = GetOutputPath(basePath, objPath);
+
+            // Remove any files in the queue that we don't care about any more
+            foreach (string file in _filesToCompile.Keys)
+            {
+                if (!inputs.Inputs.Contains(file))
+                {
+                    ImmutableInterlocked.TryRemove(ref _filesToCompile, file, out _);
+                }
+            }
+
+            _receivedProjectRuleSource.TrySetResult(true);
+
+            QueueCompilation();
+
+            void AddAllInputsToQueue(bool skipFileWriteChecks)
+            {
+                foreach (string file in inputs.Inputs)
+                {
+                    ImmutableInterlocked.TryAdd(ref _filesToCompile, file, skipFileWriteChecks);
+                }
+            }
+        }
+
+        internal static string GetOutputPath(string projectPath, string intermediateOutputPath)
+        {
+            return Path.Combine(projectPath, intermediateOutputPath, "TempPE");
+        }
+
+        private void QueueCompilation()
+        {
+            if (_filesToCompile.Count > 0)
+            {
+                _scheduler.ScheduleAsyncTask(ProcessCompileQueue, _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+            }
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                // This will stop our blocks taking any more input
+                _inputsActionBlock?.Complete();
+                _fileWatcherActionBlock?.Complete();
+
+                if (_inputsActionBlock != null)
+                {
+                    await Task.WhenAll(_inputsActionBlock.Completion, _fileWatcherActionBlock!.Completion);
+                }
+
+                // By waiting for completion we know that the following dispose will cancel any pending compilations, and there won't be any more
+                _scheduler.Dispose();
+                _disposables.Dispose();
+            }
+        }
+
+        private async Task ProcessCompileQueue(CancellationToken token)
```

Comment:
Can remove two state machines by remove async /await completely form this method and its body.

Reply:
I must be missing something, how do I remove two? Are you suggesting using a Forget for the actual compilation, so we start compiling in parallel?